pkg/apis/core/validation/validation.go

```diff
@@ -4639,17 +4655,17 @@ func validatePodAntiAffinity(podAntiAffinity *core.PodAntiAffinity, allowInvalid
 }
 
 // validateNodeAffinity tests that the specified nodeAffinity fields have valid data
-func validateNodeAffinity(na *core.NodeAffinity, fldPath *field.Path) field.ErrorList {
+func validateNodeAffinity(na *core.NodeAffinity, allowInvalidLabelValueInNodeSelector bool, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 	// TODO: Uncomment the next three lines once RequiredDuringSchedulingRequiredDuringExecution is implemented.
 	// if na.RequiredDuringSchedulingRequiredDuringExecution != nil {
 	//	allErrs = append(allErrs, ValidateNodeSelector(na.RequiredDuringSchedulingRequiredDuringExecution, fldPath.Child("requiredDuringSchedulingRequiredDuringExecution"))...)
 	// }
 	if na.RequiredDuringSchedulingIgnoredDuringExecution != nil {
-		allErrs = append(allErrs, ValidateNodeSelector(na.RequiredDuringSchedulingIgnoredDuringExecution, fldPath.Child("requiredDuringSchedulingIgnoredDuringExecution"))...)
+		allErrs = append(allErrs, ValidateNodeSelector(na.RequiredDuringSchedulingIgnoredDuringExecution, allowInvalidLabelValueInNodeSelector, fldPath.Child("requiredDuringSchedulingIgnoredDuringExecution"))...)
 	}
 	if len(na.PreferredDuringSchedulingIgnoredDuringExecution) > 0 {
-		allErrs = append(allErrs, ValidatePreferredSchedulingTerms(na.PreferredDuringSchedulingIgnoredDuringExecution, fldPath.Child("preferredDuringSchedulingIgnoredDuringExecution"))...)
+		allErrs = append(allErrs, ValidatePreferredSchedulingTerms(na.PreferredDuringSchedulingIgnoredDuringExecution, allowInvalidLabelValueInNodeSelector, fldPath.Child("preferredDuringSchedulingIgnoredDuringExecution"))...)
```

Comment:
For the preferred one, we should make a warning, instead of completely denying them, because they work even with the invalid ones (right?) unlike the required one.

That's what I interpreted from @liggitt 's answer. 
https://github.com/kubernetes/kubernetes/pull/128212#issuecomment-2435756413

Reply:
I might not have described it correctly in Slack. However, preferred affinity doesn't handle incorrect label values properly either; It returns a parse error here:
https://github.com/kubernetes/kubernetes/blob/ce47f7b416ff0e5a5c73d70e8dd3175ced5afaa9/staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go#L121-L126

I believe we need to deny incorrect label values for both types of affinity.

Reply:
I see, so NodeAffinity#Score() returns errors for invalid preferred ones and the Pods aren't working in both cases required and preferred.   Okay, got it.